<a href="https://colab.research.google.com/github/ivnlee/tec-mna-nlp/blob/main/semana-9/Equipo13_Semana09.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Maestría en Inteligencia Artificial Aplicada**
## **Curso: Procesamiento de Lenguaje Natural (NLP)**
### Tecnológico de Monterrey
### Prof. Luis Eduardo Falcón Morales
_ _ _

## **Equipo 13**

_ _ _

## **Actividad Semana 9**

### **Modelos Transformer para textos en Español e Inglés**


Importamos las librerias necesarias para la actividad

In [1]:
!pip install transformers
!pip install -q transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.2/7.2 MB 64.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 236.8/236.8 kB 36.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 39.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 40.9 MB/s eta 0:00:00


In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
import transformers
from transformers import pipeline

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


###**1. Descarga el archivo amazonbaby5000.csv que se encuentra en Canvas. Este archivo está formado de comentarios en inglés sobre productos para bebé adquiridos en la plataforma de Amazon.**

In [4]:
#Almacenamos la informacion del CSV en un Dataframe

from tensorflow.python import train
data = pd.read_csv("/content/drive/MyDrive/NLP/semana-9/amazonbaby5000.csv")

###**2. Realiza una partición de los datos en el porcentaje que consideres adecuado, entrenamiento, validación y prueba.**

In [5]:
#Realiza una partición de los datos en el porcentaje que consideres adecuado, entrenamiento,validación y prueba
#train_data, val_data, test_data = train_test_split(data["review"], data["rating"], test_size=0.2, random_state=42)

train_val_data, test_data = train_test_split(data, test_size=0.2, random_state=42)
train_data, val_data = train_test_split(train_val_data, test_size=0.25, random_state=42)

#Validamos los tamaños de los conjuntos
print("Conjunto de entrenamiento:", len(train_data))
print("Conjunto de prueba:", len(test_data))
print("Conjunto de validacion:", len(val_data))

Conjunto de entrenamiento: 3000
Conjunto de prueba: 1000
Conjunto de validacion: 1000


### **3.Utiliza un modelo Transformer de HuggingFace para análisis de sentimiento en inglés y lleva a cabo la predicción de los comentarios en los siguientes casos:**


**a.** Para utilizar un modelo Transformer predeterminado de Hugging Face de manera directa, podemos utilizar el modelo pipeline de la biblioteca transformers. Este modelo ya está pre-entrenado y listo para su uso en tareas de análisis de sentimiento en inglés.

In [6]:
!pip install datasets huggingface_hub
!pip install transformers[torch]
!pip install accelerate>=0.20.1

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 485.6/485.6 kB 33.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 110.5/110.5 kB 11.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 212.5/212.5 kB 27.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.3/134.3 kB 16.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 77.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.5/114.5 kB 20.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 268.8/268.8 kB 36.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 149.6/149.6 kB 24.0 MB/s eta 0:00:00
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 227.6/227.6 kB 16.4 MB/s eta 0:00:00


In [8]:
# Importación de librerías
from sklearn.metrics import accuracy_score
from transformers import pipeline, AutoTokenizer

# Sentiment-Analysis es un modelo de Hugging Face
model_n = "siebert/sentiment-roberta-large-english"
sentiment_pipeline = pipeline("sentiment-analysis", model=model_n, max_length=512 ,truncation=True)

# Obtener los comentarios de prueba y las etiquetas reales
test_comments = test_data["review"].tolist()
true_labels = test_data["rating"].tolist()

# Realizar la predicción de sentimiento
predictions = sentiment_pipeline(test_comments)

In [ ]:
# Obtener las etiquetas predichas
predicted_labels = [pred["label"] for pred in predictions]

# Convertir las etiquetas predichas con valores de 0 y 1 para poder comprar con los valores reales
predicted_labels_list = [1 if label == "POSITIVE" else 0 for label in predicted_labels]

# Calcular la precisión (accuracy)
accuracy = accuracy_score(true_labels, predicted_labels_list)

print("Accuracy:", accuracy)


Accuracy: 0.945


**b.** Utiliza ahora un ajuste de parámetros (fine-tuning) con el modelo Transformer DistilBERT,
para generar un nuevo modelo y clasificación de los comentarios. Es decir, el objetivo es
utilizar un modelo Transformer pre-entrenado y adaptarlo (fine-tuning) o ajustarlo a las
características de tus datos de entrenamiento. Indicar desempeño (accuracy) obtenido.

In [9]:
import torch
from transformers import AutoTokenizer, DistilBertForSequenceClassification
from torch.utils.data import DataLoader, TensorDataset
from sklearn.metrics import accuracy_score

# Cargar los conjuntos de datos de entrenamiento y validacion
train_comments = train_data["review"].tolist()  # Obtener los comentarios de entrenamiento como una lista
train_labels = train_data["rating"].tolist()  # Obtener las etiquetas de entrenamiento como una lista
test_comments = train_val_data["review"].tolist()  # Obtener los comentarios de validación como una lista
test_labels = train_val_data["rating"].tolist()  # Obtener las etiquetas de validación como una lista

# Tokenizar los comentarios utilizando el tokenizer de DistilBERT
tokenizer = AutoTokenizer.from_pretrained("distilbert-base-uncased")  # Cargar el tokenizer de DistilBERT
train_encodings = tokenizer(train_comments, truncation=True, padding=True, max_length=512)  # Tokenizar los comentarios de entrenamiento
test_encodings = tokenizer(test_comments, truncation=True, padding=True, max_length=512)  # Tokenizar los comentarios de validación

# Crear los conjuntos de datos de entrenamiento y validacion
train_dataset = TensorDataset(torch.tensor(train_encodings.input_ids),
                              torch.tensor(train_encodings.attention_mask),
                              torch.tensor(train_labels))  # Crear un conjunto de datos para el entrenamiento
test_dataset = TensorDataset(torch.tensor(test_encodings.input_ids),
                             torch.tensor(test_encodings.attention_mask),
                             torch.tensor(test_labels))  # Crear un conjunto de datos para la validación

# Cargar el modelo pre-entrenado de DistilBERT
model = DistilBertForSequenceClassification.from_pretrained("distilbert-base-uncased", num_labels=2)  # Cargar el modelo pre-entrenado de DistilBERT con 2 etiquetas

# Entrenamiento del modelo
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")  # Verificar si hay una GPU disponible
model.to(device)  # Mover el modelo a la GPU si está disponible, de lo contrario, se utilizará la CPU
train_loader = DataLoader(train_dataset, batch_size=8, shuffle=True)  # Cargar el conjunto de datos de entrenamiento en lotes
optimizer = torch.optim.Adam(model.parameters(), lr=1e-5)  # Definir el optimizador Adam con una tasa de aprendizaje de 1e-5
loss_fn = torch.nn.CrossEntropyLoss()  # Definir la función de pérdida de entropía cruzada

model.train()  # Establecer el modo de entrenamiento del modelo
for epoch in range(3):  # Ejecutar el entrenamiento durante 3 épocas
    total_loss = 0
    for batch in train_loader:
        batch = tuple(t.to(device) for t in batch)
        input_ids, attention_mask, labels = batch

        optimizer.zero_grad()  # Reiniciar los gradientes acumulados en cada iteración
        outputs = model(input_ids=input_ids, attention_mask=attention_mask, labels=labels)  # Obtener las salidas del modelo y calcular la pérdida
        loss = outputs.loss  # Obtener la pérdida
        total_loss += loss.item()  # Acumular la pérdida total

        loss.backward()  # Realizar la retropropagación para calcular los gradientes
        optimizer.step()  # Actualizar los parámetros del modelo utilizando el optimizador

    avg_loss = total_loss / len(train_loader)  # Calcular la pérdida promedio
    print(f"Epoch {epoch+1}/{3}, Loss: {avg_loss}")  # Imprimir la pérdida promedio de la época actual

# Evaluación del modelo
model.eval()  # Establecer el modo de evaluación del modelo
test_loader = DataLoader(test_dataset, batch_size=8)  # Cargar el conjunto de datos de validación en lotes
predicted_labels = []  # Lista para almacenar las etiquetas predichas
true_labels = []  # Lista para almacenar las etiquetas verdaderas

with torch.no_grad():
    for batch in test_loader:
        batch = tuple(t.to(device) for t in batch)
        input_ids, attention_mask, labels = batch

        outputs = model(input_ids=input_ids, attention_mask=attention_mask)  # Obtener las salidas del modelo sin calcular la pérdida
        logits = outputs.logits  # Obtener los logits (valores de salida antes de aplicar la función softmax)
        predicted_labels.extend(torch.argmax(logits, dim=1).cpu().tolist())  # Obtener las etiquetas predichas y agregarlas a la lista
        true_labels.extend(labels.cpu().tolist())  # Obtener las etiquetas verdaderas y agregarlas a la lista

accuracy = accuracy_score(true_labels, predicted_labels)  # Calcular la precisión del modelo comparando las etiquetas verdaderas y predichas
print("Accuracy:", accuracy)  # Imprimir la precisión del modelo


Some weights of the model checkpoint at distilbert-base-uncased were not used when initializing DistilBertForSequenceClassification: ['vocab_transform.bias', 'vocab_projector.bias', 'vocab_transform.weight', 'vocab_layer_norm.weight', 'vocab_layer_norm.bias']
- This IS expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.weight', 'pre_classifier.weight', 'classifier.bias', 'pre_classifier.

Epoch 1/3, Loss: 0.36119764108459157
Epoch 2/3, Loss: 0.1902855358024438
Epoch 3/3, Loss: 0.1012801325644056
Accuracy: 0.973


#### **C.** Compara los resultados e incluye tus comentarios.

Antes de comparar los resultados vale la pena mencionar que el primer modelo utilizado ROBERTA, fue desarrollada partiendo de BERT por lo que en teoria es una versión mejorada, mientras que el segundo modelo DISTILBERT es una version "compromida" de BERT pues cuenta con menos parametros y ha sido entrenada con un conjunto de datos mas pequeño.

Con Roberta obtuvimos como resultado 94% de precision mientras que con Distilbert obtuvimos 97% de precisión, aunque aparentemente en el segundo modelo obtuvimos un mejor resultado, este fue conseguido en el 3er epoch despues de aplicar fine-tuning; Mientras que, el resultado de Roberta se obtuvo al utilizar el modelo sin realizar ningun ajuste a los hiperparametros.

Con estos resultados podemos deducir que de haber ajustado el modelo 1 del mismo modo que se hizo con el 2, probablemente Roberta hubiera conseguido un mejor resultado, aún asi, comparando las primeras ejecuciones notamos que el gracias a la arquitectura del modelo y su los datos con los que fue entrenado obtiene un mejor modelo considerablemente.



#### **4. Utiliza el modelo Transformer pre-entrenado para diferentes idiomas llamado Helsinki-NLP de HF,para hacer una traducción del Inglés al Español de los 5000 comentarios. Observa algunos de los comentarios traducidos y comenta qué tan bien consideras dichas traducciones. Por el momento es un simple comentario subjetivo, aunque existen métricas y bases de datos para evaluar el desempeño de un Trasnformer**

In [ ]:
!pip install sentencepiece
!pip install --upgrade transformers
!pip install sacremoses

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
from transformers import MarianTokenizer, MarianMTModel, pipeline

# Importamos el modelo y el tokenizador
model_name = "Helsinki-NLP/opus-mt-en-es"
tokenizer = MarianTokenizer.from_pretrained(model_name)
model = MarianMTModel.from_pretrained(model_name)

# Creamos el pipeline de traducción
translate_pipe = pipeline('translation', model=model, tokenizer=tokenizer, truncation=True, batch_size=8, skip_special_tokens=True)

# Guardamos los datos de entrenamiento en una variable
english_comments = data["review"].tolist()[:5000]

# Realizamos la traducción
translated_comments = translate_pipe(english_comments, max_length=512)

# Decodificamos las traducciones e imprimimos los primeros 5 comentarios
for i in range(5):
    print("Comentario original (inglés):", english_comments[i])
    print("Comentario traducido (español):", translated_comments[i]['translation_text'])
    print()


In [ ]:
for i in range(5):
    print("Comentario original (inglés):", english_comments[i])
    print("Comentario traducido (español):", translated_comments[i])
    print()

###**5. Utiliza el modelo Transformer “pysentimiento”, basado en BETO y este a su vez basado en BERT, para llevar a cabo un análisis de sentimiento (sentiment analysis) y hacer la predicción de los comentarios traducidos al español. Incluye tus comentarios del esultado obtenido.**

In [ ]:
translated_comments = pd.read_csv("amazonbaby5000-es.csv")

#Realiza una partición de los datos en el porcentaje que consideres adecuado, entrenamiento,validación y prueba
#train_data, val_data, test_data = train_test_split(data["review"], data["rating"], test_size=0.2, random_state=42)

train_val_data_es, test_data_es = train_test_split(translated_comments, test_size=0.2, random_state=42)
train_data_es, val_data_es = train_test_split(train_val_data_es, test_size=0.25, random_state=42)

#Validamos los tamaños de los conjuntos
print("Conjunto de entrenamiento:", len(train_data_es))
print("Conjunto de prueba:", len(test_data_es))
print("Conjunto de validacion:", len(val_data_es))

Conjunto de entrenamiento: 3000
Conjunto de prueba: 1000
Conjunto de validacion: 1000


In [ ]:
!pip install transformers
!wget https://users.dcc.uchile.cl/~jperez/beto/cased_2M/pytorch_weights.tar.gz
!wget https://users.dcc.uchile.cl/~jperez/beto/cased_2M/vocab.txt
!wget https://users.dcc.uchile.cl/~jperez/beto/cased_2M/config.json
!tar -xzvf pytorch_weights.tar.gz
!mv config.json pytorch/.
!mv vocab.txt pytorch/.

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
--2023-06-21 23:26:48--  https://users.dcc.uchile.cl/~jperez/beto/cased_2M/pytorch_weights.tar.gz
Resolving users.dcc.uchile.cl (users.dcc.uchile.cl)... 200.9.99.211, 192.80.24.4
Connecting to users.dcc.uchile.cl (users.dcc.uchile.cl)|200.9.99.211|:443... connected.
HTTP request sent, awaiting response... 403 Forbidden
2023-06-21 23:26:49 ERROR 403: Forbidden.

--2023-06-21 23:26:49--  https://users.dcc.uchile.cl/~jperez/beto/cased_2M/vocab.txt
Resolving users.dcc.uchile.cl (users.dcc.uchile.cl)... 200.9.99.211, 192.80.24.4
Connecting to users.dcc.uchile.cl (users.dcc.uchile.cl)|200.9.99.211|:443... connected.
HTTP request sent, awaiting response... 403 Forbidden
2023-06-21 23:26:50 ERROR 403: Forbidden.

--2023-06-21 23:26:50--  https://users.dcc.uchile.cl/~jperez/beto/cased_2M/config.json
Resolving users.dcc.uchile.cl (users.dcc.uchile.cl)... 200.9.99.211, 192.80.24.4
Connecting to user

In [ ]:
# Importación de librerías
import torch
from transformers import BertForMaskedLM, BertTokenizer

# Crea el pipeline de análisis de sentimiento
model_es = "dccuchile/bert-base-spanish-wwm-uncased"
sentiment_analysis = pipeline("sentiment-analysis", model=model_es, tokenizer=model_es)


# Tokenizar los comentarios
predictions = sentiment_analysis(test_data_es["reviews"].tolist())

Some weights of the model checkpoint at dccuchile/bert-base-spanish-wwm-uncased were not used when initializing BertForSequenceClassification: ['cls.predictions.decoder.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.decoder.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.bias', 'cls.predictions.transform.LayerNorm.weight']
- This IS expected if you are initializing BertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at dccuc

Xformers is not installed correctly. If you want to use memory_efficient_attention to accelerate training use the following command to install Xformers
pip install xformers.


In [ ]:
from sklearn.metrics import accuracy_score
from transformers import pipeline
from transformers import BertForSequenceClassification


# Cargamos el modelo y el tokenizer de BERT en español
model_name = "dccuchile/bert-base-spanish-wwm-uncased"
tokenizer = BertTokenizer.from_pretrained(model_name)
model = BertForSequenceClassification.from_pretrained(model_name)

# Comentarios en español
text_list = translated_comments["reviews"].tolist()# Lista de comentarios en español
label_list = translated_comments["labels"].tolist()# Lista de etiquetas de sentimiento reales (0 para negativo, 1 para positivo)
predicted_labels = []

# Usamos el modelo y tokenizer para hacer el análisis de sentimientos cada comentario
for text, label in zip(text_list, label_list):
    encoded_input = tokenizer(text, padding=True, truncation=True, return_tensors="pt")
    logits = model(**encoded_input).logits
    predicted_class = torch.argmax(logits, dim=1).item()
    predicted_labels.append(predicted_class)

# Calculamos la precisión que obtuvimos
accuracy = accuracy_score(label_list, predicted_labels)
print(f"Accuracy: {accuracy}")

Some weights of the model checkpoint at dccuchile/bert-base-spanish-wwm-uncased were not used when initializing BertForSequenceClassification: ['cls.predictions.decoder.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.decoder.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.bias', 'cls.predictions.transform.LayerNorm.weight']
- This IS expected if you are initializing BertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at dccuc

Accuracy: 0.6922


Como se observa, se está acertando el sentimiento en el 62% de los comentarios, para intentar mejorar el acierto, vamos a realizar algunos pasos más de pre-procesamiento.

In [ ]:
import re
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem import SnowballStemmer

# Vamos a utilizar stemmer y obtener las stopwords en español
stemmer = SnowballStemmer("spanish")
stopwords_es = set(stopwords.words("spanish"))

# Creamos una función para aplicar el preprocesamiento
def preprocess_text(text):
    # se convierte todo a minúsculas
    text = text.lower()
    # removemos los caracteres especiales y puntuación
    text = re.sub(r"[^a-zA-Z0-9áéíóúüñ]", " ", text)
    # Tokenizamos
    tokens = text.split()
    # Se eliminan las stopwords
    tokens = [token for token in tokens if token not in stopwords_es]
    # Aplicamos stemming
    tokens = [stemmer.stem(token) for token in tokens]
    # unimos los tokens nuevamente
    preprocessed_text = " ".join(tokens)
    return preprocessed_text

# Generamos las listas con las etiquetas observadas y el nuevo conjunto ya preprocesado
label_list = translated_comments["labels"].tolist()
preprocessed_text_list = [preprocess_text(text) for text in translated_comments["reviews"].tolist()]

predicted_labels = []

# Usamos el modelo y tokenizer para hacer el análisis de sentimientos cada comentario
for text in preprocessed_text_list:
    encoded_input = tokenizer(text, padding=True, truncation=True, return_tensors="pt")
    logits = model(**encoded_input).logits
    predicted_class = torch.argmax(logits, dim=1).item()
    predicted_labels.append(predicted_class)

# Calculamos la precisión que obtuvimos
accuracy = accuracy_score(label_list, predicted_labels)
print(f"Accuracy: {accuracy}")

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


Accuracy: 0.7654


Con el preprocesamiento de los datos obtuvimos una mejoría importante, ahora el modelo está acertando el sentimiento del 75% de los comentarios.

### **6. Resume los resultados obtenidos en esta actividad de manera tabular e incluye tus comentarios finales.**

| Inciso | Modelo | Fine-tuning | Accuracy | Comentarios: |
| :---:  | :----: | :---------: | :------: | :---------- |
| 3a. | RoBERTA | No | 94.5 %| Clasificación de texto en inglés |
| 3b. | DistilBERT | Si | 97.3 %| Clasificación de texto en inglés |
| 5 (1)  | BETO | No | 69.2 %| Clasificación del texto traducido al español |
| 5 (2)  | BETO | No | 76.5 %| Clasificación del texto traducido al español con pre-procesamiento |